### PR TITLE
New semconvgen templates prototype

### PR DESCRIPTION
**Pull request by lmolkova**
Prototype for new semconv generation approach (that groups attributes by namespace and stability vs semconv)
https://github.com/open-telemetry/semantic-conventions/issues/551

## Changes

- Breaks down attribute definitions into individual files based on their root namespace
- Defines experimental attributes in the corresponding `vX_Y_Z_Experimental` namespace
- Demonstrates how metric definitions can be auto-generated

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/CloudeventsAttributes.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// <auto-generated> This file has been auto generated from buildscripts/semantic-conventions/templates/Attributes.cs.j2</auto-generated>
+
+using System;
+
+namespace OpenTelemetry.SemanticConventions.v1_23_1_Experimental
+{
+    /// <summary>
+    /// Describes semantic conventions for attributes in the <c>cloudevents</c> namespace.
+    /// </summary>
+    public static class CloudeventsAttributes
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/CloudeventsAttributes.cs:26]: `cloudevents`, `graphql`, `webengine`, `opentracing`, etc cannot be automatically CamelCased, so need to figure out something on the semconv side to fix it across languages

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/ClientAttributes.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    /// </summary>
+    public static class ClientAttributes
+    {
+        /// <summary>
+        /// Client address - domain name if available without reverse DNS lookup; otherwise, IP address or Unix domain socket name.
+        /// </summary>
+        /// <remarks>
+        /// When observed from the server side, and when communicating through an intermediary, <c>client.address</c> SHOULD represent the client address behind any intermediaries,  for example proxies, if it&amp;#39;s available.
+        /// </remarks>
+        public const string ClientAddress = "client.address";
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventions/ClientAttributes.cs:34]: Q:
`ClientAttributes.ClientAddress` vs `ClientAttributes.Address` ?

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/OtherAttributes.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// <auto-generated> This file has been auto generated from buildscripts/semantic-conventions/templates/Attributes.cs.j2</auto-generated>
+
+using System;
+
+namespace OpenTelemetry.SemanticConventions.v1_23_1_Experimental
+{
+    /// <summary>
+    /// Describes semantic conventions for attributes in the <c>other</c> namespace.
+    /// </summary>
+    public static class OtherAttributes
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/OtherAttributes.cs:26]: todo on the spec side: all attributes must have a namespace

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/HttpMetrics.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  /// </summary>
+  public static class HttpMetrics {
+      /// <summary>
+      /// Creates <c>http.client.request.body.size</c> instrument.
+      /// Size of HTTP client request bodies.
+      /// </summary>
+      /// <remarks>
+      /// The size of the request payload body in bytes. This is the number of bytes transferred excluding headers and is often, but not always, present as the <a href="https://www.rfc-editor.org/rfc/rfc9110.html#field.content-length">Content-Length</a> header. For requests using transport encoding, this should be the compressed size.
+      /// </remarks>
+      public static Histogram<double> CreateHttpClientRequestBodySize(Meter meter)
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/HttpMetrics.cs:34]: value type is not part of yaml - https://github.com/open-telemetry/semantic-conventions/issues/591, so it's not fully possible yet

File: src/OpenTelemetry.SemanticConventions/scripts/semantic-conventions/generate.ps1 [hunk 5]
```diff
@@ -3,47 +3,64 @@ $SCRIPT_DIR=$PSScriptRoot
...
+  --output /output/Attributes.cs `
   --trim-whitespace `
-  -D class=ResourceSemanticConventions `
-  -D schemaUrl=$SCHEMA_URL `
-  -D pkg=OpenTelemetry.Resources
+  --file-per-group root_namespace `
+  -D filter=is_experimental `
+  -D pkg=OpenTelemetry.SemanticConventions.${SPEC_VERSION_ESCAPED}
+
+#experimental metrics
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/semantic-conventions/generate.ps1:51]: no stable metric yet - https://github.com/open-telemetry/semantic-conventions/pull/587

File: src/OpenTelemetry.SemanticConventions/scripts/semantic-conventions/generate.ps1 [hunk 6]
```diff
@@ -3,47 +3,64 @@ $SCRIPT_DIR=$PSScriptRoot
...
+  --output /output/Attributes.cs `
   --trim-whitespace `
-  -D class=TraceSemanticConventions `
-  -D schemaUrl=$SCHEMA_URL `
-  -D pkg=OpenTelemetry.Trace
+  --file-per-group root_namespace `
+  -D filter=is_stable `
+  -D pkg=OpenTelemetry.SemanticConventions
 
+#experimental attributes
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/scripts/semantic-conventions/generate.ps1:37]: it's easy to filter stable/experimental and move them into different artifacts, use different templates, add `[Experimental]` attribute, etc

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/CloudeventsAttributes.cs (hunk 1 above)
Comment (lmolkova) [reply src/OpenTelemetry.SemanticConventions/SemanticConventions/v1_23_1_Experimental/CloudeventsAttributes.cs:26]: https://github.com/open-telemetry/semantic-conventions/issues/599

File: src/OpenTelemetry.SemanticConventions/SemanticConventions/HttpAttributes.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        /// the list of known HTTP methods. If this override is done via environment variable, then the environment variable MUST be named
+        /// OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS and support a comma-separated list of case-sensitive known HTTP methods
+        /// (this list MUST be a full override of the default known method, it is not a list of known methods in addition to the defaults).HTTP method names are case-sensitive and <c>http.request.method</c> attribute value MUST match a known HTTP method name exactly.
+        /// Instrumentations for specific web frameworks that consider HTTP methods to be case insensitive, SHOULD populate a canonical equivalent.
+        /// Tracing instrumentations that do so, MUST also set <c>http.request.method_original</c> to the original value.
+        /// </remarks>
+        public const string HttpRequestMethod = "http.request.method";
```
Comment (lmolkova) [inline src/OpenTelemetry.SemanticConventions/SemanticConventions/HttpAttributes.cs:51]: Would we want multiple layers like `HttpAttributes.Request`?

**Comment by martinjt** [comment]
I'm slightly worried about the idea that "use whatever is the latest", versus pinning to a specific version.

The problem I see is that if I use:

`OpenTelemetry.SemanticConventions.HttpAttributes.HttpRequestMethod` and a V2 of the semantic conventions changes that, I'm going to automatically inherit that change in my library if the consumer is using a V2 version.

I think having the ability to have `V1_23_1` namespace, and the default you've provided allows authors to opt-in to "whatever is new", or "this specific version of the conventions".

(missed the SIG, sorry if it was discussed there)

**Pull request closed without merging**